### ISSUE TITLE
Peakfinder finds double peaks 

### ISSUE THREAD

**Issue by arunpersaud** [lbl-anp/becquerel#263]
Hello

when using a LaBr spectrum (I assume you mostly use Ge?) I get double peaks from the peakfinding algorithm. The problem seems to be in 

d2 = self.snr[2:] - 2 * self.snr[1:-1] + self.snr[:-2]

of peakfinder.py

![Figure 1](https://user-images.githubusercontent.com/425653/114461802-08817180-9b97-11eb-99ed-281af41ee2b7.png)

if I change the above line to

peak = (d1[2:] < 0) & (d1[1:-1] > 0) & (d2[1:-1] < 0)

The issue seems to be resolved.

![Figure 1(2)](https://user-images.githubusercontent.com/425653/114461946-3f578780-9b97-11eb-9abb-3f17fbd1e7da.png)

Let me know what you think and I can create a PR for this.

**Comment by jvavrek** [comment]
@arunpersaud I have noticed the same behavior recently in CLLBC. Thanks for the proposed fix—I'll take a look when I have a chance!

**Comment by arunpersaud** [comment]
sounds good... this is around line 387 in core/peakfinder.py

**Comment by jvavrek** [comment]
@arunpersaud do you still have this issue after the recent PRs? I don't have this problem in my analysis anymore, but that may be a result of adjusting the FWHM until the problem disappeared.

**Comment by arunpersaud** [comment]
yes, I had this problem still after fixing the FWHM

**Comment by jvavrek** [comment]
@arunpersaud I was able to replicate this. With your snippet above, on my data, it picks the left point whereas the right point should be the identified peak. @markbandstra and I are working on a fix in #266.

**Comment by jvavrek** [comment]
@arunpersaud can you test if PR #266 works for you? If so, can we add your example above as a test?

**Comment by markbandstra** [comment]
I was able to reproduce this behavior as well on some test data, but only if I pulled out the [specific code in peakfinder.py](https://github.com/lbl-anp/becquerel/blob/main/becquerel/core/peakfinder.py#L393-L402) that calculates the first and second derivatives and the peak finding logic, like this:

```python
x = np.arange(31)
y = 1.5 * np.exp(-0.5 * (x[:-1] - 15) ** 2 / (2 * 3 ** 2))
spec = bq.Spectrum(bin_edges_raw=x, counts=y)
kernel = bq.GaussianPeakFilter(15, 7.5, fwhm_at_0=7.5)
finder = bq.PeakFinder(spec, kernel)

# code from PeakFinder.find_peaks
d1 = (finder.snr[2:] - finder.snr[:-2]) / 2
d1 = np.append(0, d1)
d1 = np.append(d1, 0)
d2 = finder.snr[2:] - 2 * finder.snr[1:-1] + finder.snr[:-2]
d2 = np.append(0, d2)
d2 = np.append(d2, 0)
peak = (d1[2:] < 0) & (d1[:-2] > 0) & (d2[1:-1] < 0)
peak = np.append(False, peak)
peak = np.append(peak, False)

# double peaks are apparent here
plt.figure()
plt.plot(y, color="k", label="raw data")
plt.plot(d1, label="d1")
plt.plot(d2, label="d2")
plt.plot(finder.snr, label="SNR")
plt.plot(peak, marker=".", label="peaks")
plt.legend()
plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/1197880/117048010-5fc2bf80-acc7-11eb-8978-7c19b1e5718e.png)

However, the peak centroids that are found by this code are then sent to [`PeakFinder.add_peak`](https://github.com/lbl-anp/becquerel/blob/main/becquerel/core/peakfinder.py#L249-L300), which checks whether the centroid has already been recorded by [ensuring any new centroid is at least `min_sep` channels away from any others](https://github.com/lbl-anp/becquerel/blob/main/becquerel/core/peakfinder.py#L260-L263):

```python
        is_new_x = True
        for cent in self.centroids:
            if abs(xpeak - cent) <= self.min_sep:
                is_new_x = False
```

And the double peak does not show up in the list of centroids:

```python
# double peaks are not apparent here
finder.find_peaks(min_snr=1.0)
print(finder.centroids)

# Out: [14.5]
```

@jvavrek and @arunpersaud when you see this issue, is it actually making it to `PeakFinder.centroids` or is this code working as intended to avoid duplicates?

**Comment by jvavrek** [comment]
@markbandstra great point, I think I only see this problem when I don't specify `min_sep` appropriately.

Maybe we can default `min_sep` to the bin width?

**Comment by markbandstra** [comment]
@jvavrek sounds good, and I'm also in favor of adjusting the peak-finding logic as you've already done on #266.

Maybe something like having `min_sep=None` as default, and then if it is `None` use the `spectrum` to find the bin width.

**Comment by jvavrek** [comment]
@markbandstra I agree. The only issue is non-uniform bins. I guess we can look up the bin width for each trial centroid.

Re: #266, I just adapted your code above into a regression test of sorts. The old logic, written directly into the test, will produce two peaks, while the new logic, called through `finder.find_peaks()`, is tested to ensure only one peak is produced.

**Comment by jvavrek** [comment]
@arunpersaud we just merged #266 . Can you check if that works for your data?

**Comment by jvavrek** [comment]
@arunpersaud closing for now. Let us know if you run into this issue again.
